p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,465 @@
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not generate dial args: %w", err)
+	}
+
+	raddr, err := net.ResolveUDPAddr(rnw, rhost)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not resolve udp address: %w", err)
+	}
+
+	// Instead of encoding the local fingerprint we
+	// instead generate a random uuid as the connection ufrag.
+	// The only requirement here is that the ufrag and password
+	// must be equal, which will allow the server to determine
+	// the password using the STUN message.
+	ufrag := "libp2p+webrtc+v1/" + genUfrag(32)
+
+	settingEngine := webrtc.SettingEngine{}
+	// suppress pion logs
+	loggerFactory := pionlogger.NewDefaultLoggerFactory()
+	loggerFactory.DefaultLogLevel = pionlogger.LogLevelDisabled
+	settingEngine.LoggerFactory = loggerFactory
+
+	settingEngine.SetICECredentials(ufrag, ufrag)
+	settingEngine.DetachDataChannels()
+	settingEngine.SetAnsweringDTLSRole(webrtc.DTLSRoleClient)
+	settingEngine.SetIncludeLoopbackCandidate(true)
+	settingEngine.SetICETimeouts(t.peerConnectionDisconnectedTimeout, t.peerConnectionFailedTimeout, t.peerConnectionKeepaliveTimeout)
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(settingEngine))
+
+	pc, err = api.NewPeerConnection(t.webrtcConfig)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not instantiate peerconnection: %w", err)
+	}
+
+	errC := awaitPeerConnectionOpen(ufrag, pc)
+	// We need to set negotiated = true for this channel on both
+	// the client and server to avoid DCEP errors.
+	negotiated, id := hansdhakeChannelNegotiated, handshakeChannelId
+	rawHandshakeChannel, err := pc.CreateDataChannel("", &webrtc.DataChannelInit{
+		Negotiated: &negotiated,
+		ID:         &id,
+	})
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not create datachannel: %w", err)
+	}
+
+	// do offer-answer exchange
+	offer, err := pc.CreateOffer(nil)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not create offer: %w", err)
+	}
+
+	err = pc.SetLocalDescription(offer)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not set local description: %w", err)
+	}
+
+	answerSdpString := renderServerSdp(raddr, ufrag, remoteMultihash)
+
+	answer := webrtc.SessionDescription{SDP: answerSdpString, Type: webrtc.SDPTypeAnswer}
+	err = pc.SetRemoteDescription(answer)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not set remote description: %w", err)
+	}
+
+	// await peerconnection opening
+	select {
+	case err := <-errC:
+		if err != nil {
+			return pc, nil, err
+		}
+	case <-ctx.Done():
+		return pc, nil, fmt.Errorf("peerconnection opening timed out")
+	}
+
+	detached, err := getDetachedChannel(ctx, rawHandshakeChannel)
+	if err != nil {
+		return pc, nil, err
+	}
+	// set the local address from the candidate pair
+	cp, err := rawHandshakeChannel.Transport().Transport().ICETransport().GetSelectedCandidatePair()
+	if cp == nil || err != nil {
+		return pc, nil, fmt.Errorf("ice connection did not have selected candidate pair")
```

Comment:
might want to split up these cases, as now this error gets surpresed, meaning it will also never be logged

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)